src/types/step.d.ts

```diff
@@ -0,0 +1,241 @@
+        scrollToHandler: (() => any);//TODO superheri What is the return type of this? Void?
+
+        /**
+         * Should a cancel “✕” be shown in the header of the step?
+         */
+        showCancelLink: boolean;
+
+        /**
+         * A function that, when it returns `true`, will show the step.
+         * If it returns `false`, the step will be skipped.
+         */
+        showOn: (() => boolean);
+
+        /**
+         * The text in the body of the step. It can be one of four types:
+         * ```
+         * - HTML string
+         * - Array of HTML strings
+         * - `HTMLElement` object
+         * - `Function` to be executed when the step is built. It must return one of the three options above.
+         * ```
+         */
+        text: string | ReadonlyArray<string> | HTMLElement | (() => string | ReadonlyArray<string> | HTMLElement)
+
+        /**
+         * The step's title. It becomes an `h3` at the top of the step.
+         */
+        title: string;
+
+        /**
+         * You can define `show`, `hide`, etc events inside `when`. For example:
+         * ```js
+         * when: {
+         *   show: function() {
+         *     window.scrollTo(0, 0);
+         *   }
+         * }
+         * ```
+         */
+        when: StepOptionsWhen;
+    }
+
+    interface StepOptionsAttachTo {
+        /**
+         * TODO superheri add comment explaining?
+         */
+        element: HTMLElement | string;
+
+        /**
+         * TODO superheri add comment explaining?
+         */
+        on: string;
+    }
+
+    interface StepOptionsAdvanceOn {
+        selector: string;
+        event: string;
+    }
+
+    interface StepOptionsButton {
+        /**
+         * A function executed when the button is clicked on
+         */
+        action: (() => any);//TODO superheri Return type
+
+        /**
+         * Extra classes to apply to the `<a>`
+         */
+        classes: string;
+
+        /**
+         * A hash of events to bind onto the button, for example
+         * `{'mouseover': function(){}}`. Adding a `click` event to events when you already have an `action` specified is not supported.
+         * You can use events to skip steps or navigate to specific steps, with something like:
+         * ```js
+         * events: {
+         *   click: function() {
+         *     return Shepherd.activeTour.show('some_step_name');
+         *   }
+         * }
+         * ```
+         */
+        events: StepOptoonsButtonEvent;
+
+        /**
+         * The HTML text of the button
+         */
+        text: string;
+    }
+
+    interface StepOptoonsButtonEvent {
+        [key: string]: (() => void);
+    }
+
+    /**
+     * TODO superheri use better naming and comments for sub objects?
+     */
+    interface StepOptionsWhen {
+        [key: string]: (() => void);
+    }
+}
+
+/**
+ * Class representing steps to be added to a tour
+ * @extends {Evented}
+ */
+declare class Step extends Evented {
```

Comment:
again, not clear why these would be split? https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html#classes

Reply:
The reason I split the class we want to export and the sub types is that what we want to export is mostly the Step class. The other types are used to add types for the different properties of the main exported class. I based this on the explanations I received from a guy from Microsoft in the DefinitelyTyped official repo when I created type definitions for an other package : https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34714

Reply:
Here is an example from TypeScript : https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-class-d-ts.html